edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import http
+import json
+import urllib.parse
+import httpx
+import datetime
+import base64
+import immutables
+import httptools
+
+from typing import Any, Type, TypeVar, TYPE_CHECKING
+from jwcrypto import jwt, jwk
+
+from edb import errors
+from edb.common import debug
+from edb.common import markup
+from edb.server.config import ops
+
+
+# TODO get this from edb.server.config.ops directly?
+SettingsMap = immutables.Map[str, ops.SettingValue]
+
+
+# Base class for OAuth 2 Providers
+class BaseProvider:
+    def __init__(self, name: str, client_id: str, client_secret: str):
+        self.name = name
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+    def get_code_url(self) -> str:
+        raise NotImplementedError
+
+    async def exchange_code(self, code: str) -> str:
+        raise NotImplementedError
+
+    async def fetch_user_info(self, token: str) -> dict:
+        raise NotImplementedError
+
+
+# OAuth 2 Client HTTP Endpoints
+
+
+def redirect_to_auth_provider(
+    response,
+    redirect_uri: str,
+    iss: str,
+    provider_name: str,
+    db_config: SettingsMap,
+):
+    client_id, client_secret = _get_client_credentials(provider_name, db_config)
+    provider = _get_provider(
+        name=provider_name,
+        client_id=client_id,
+        client_secret=client_secret,
+    )
+
+    signing_key = _get_auth_signing_key(db_config)
+    state_token = _make_signed_token(
+        iss=iss, provider=provider_name, key=signing_key
+    )
+
+    auth_url = provider.get_code_url(
+        state=state_token.serialize(), redirect_uri=redirect_uri
+    )
+
+    response.status = http.HTTPStatus.FOUND
+    response.custom_headers["Location"] = auth_url
+    response.close_connection = True
+
+
+async def handle_auth_callback(request, response):
+    return
+
+
+async def handle_request(request, response, db, args):
+    try:
+        # Set up routing to the appropriate handler
+        if args[0] == "authorize":
+            provider_name: str | None = _get_search_param(
+                request.url.query.decode("ascii"), "provider"
+            )
+            if provider_name is None:
+                raise errors.InternalServerError("No provider specified")
```

Comment:
ISE are reserved only for unhandled server errors, they must never leak to an API or it's a bug.

You can find the full list of errors here: https://github.com/edgedb/edgedb/blob/master/edb/api/errors.txt

I'm not actually sure what error types we should use for auth. Maybe stick with BackendError for now...

Reply:
Ok, I'll switch these all to `BackendError` pending more details about how errors ought to work for extensions like this.